### PR TITLE
add WorkManager extension

### PR DESCRIPTION
**Pull request by loki666**
implementation of Scheduler using Android Arch WorkManager

### CODE REVIEW

**Comment by loki666** [comment]
I signed it!

**Comment by ojw28** [comment]
Would you see this as replacing the `jobdispatcher` extension, or are there use cases for both?

**Comment by loki666** [comment]
it can be a replacement, since WorkManager will either use platform JobDispatcher, Firebase/PlayStore jobDispatcher implementation or AlertManager depending on the API level or simple a background thread if the app is in the foreground

File: extensions/workmanager/README.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# ExoPlayer WorkManager extension #
+
+This extension provides a Scheduler implementation which uses [Android Arch WorkManager][].
+
+[Android Arch WorkManager]: https://developer.android.com/topic/libraries/architecture/workmanager.html
+
+## Getting the extension ##
+
+The easiest way to use the extension is to add it as a gradle dependency:
```
Comment (piotrek1543) [inline extensions/workmanager/README.md]: I am not grammar nazi, but `gradle` in both cases should be capitalized

**Comment by loki666** [comment]
is something missing in this PR ?

**Comment by ojw28** [comment]
Nothing is missing. We're just waiting for a convenient time to look at & merge this. Thanks!

File: extensions/workmanager/README.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# ExoPlayer WorkManager extension #
+
+This extension provides a Scheduler implementation which uses [Android Arch WorkManager][].
```
Comment (erdemguven) [inline extensions/workmanager/README.md]: Please replace with just "WorkManager" here and elsewhere.

File: extensions/workmanager/build.gradle [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    testOptions.unitTests.includeAndroidResources = true
+}
+
+dependencies {
+    implementation project(modulePrefix + 'library-core')
+    implementation 'android.arch.work:work-runtime:1.0.0'
+}
+
+ext {
+    javadocTitle = 'Android Arch WorkManager extension'
```
Comment (erdemguven) [inline extensions/workmanager/build.gradle]: "WorkManager extension"

File: extensions/workmanager/src/main/java/com/google/android/exoplayer2/ext/workmanager/WorkManagerScheduler.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+import androidx.work.Constraints;
+import androidx.work.Data;
+import androidx.work.ExistingWorkPolicy;
+import androidx.work.NetworkType;
+import androidx.work.OneTimeWorkRequest;
+import androidx.work.WorkManager;
+import androidx.work.Worker;
+import androidx.work.WorkerParameters;
+
+/***
```
Comment (erdemguven) [inline extensions/workmanager/src/main/java/com/google/android/exoplayer2/ext/workmanager/WorkManagerScheduler.java]: /**

File: extensions/workmanager/src/main/java/com/google/android/exoplayer2/ext/workmanager/WorkManagerScheduler.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                break;
+            default:
+                throw new UnsupportedOperationException();
+        }
+
+        if (requirements.isChargingRequired()) {
+            builder.setRequiresCharging(true);
+        }
+
+        if (requirements.isIdleRequired() && Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
```
Comment (erdemguven) [inline extensions/workmanager/src/main/java/com/google/android/exoplayer2/ext/workmanager/WorkManagerScheduler.java]: Why version check is necessary?

**Review by erdemguven — COMMENTED** [review]
LGTM, please fix/respond pointed issues.

File: extensions/workmanager/src/main/java/com/google/android/exoplayer2/ext/workmanager/WorkManagerScheduler.java (hunk 5 above)
Comment (loki666) [reply extensions/workmanager/src/main/java/com/google/android/exoplayer2/ext/workmanager/WorkManagerScheduler.java]: thats how the Builder API is defined
`        @RequiresApi(23)
        public @NonNull Builder setRequiresDeviceIdle(boolean requiresDeviceIdle) {
            this.mRequiresDeviceIdle = requiresDeviceIdle;
            return this;
        }`

**Comment by loki666** [comment]
since dev-v2 migrated to androidX, should I also change the dependency?
for androidX it should be 'androidx.work:work-runtime:2.0.1'

File: extensions/workmanager/src/main/java/com/google/android/exoplayer2/ext/workmanager/WorkManagerScheduler.java (hunk 5 above)
Comment (erdemguven) [reply extensions/workmanager/src/main/java/com/google/android/exoplayer2/ext/workmanager/WorkManagerScheduler.java]: Ahh, true, I checked the wrong method.
Please replace it with:
Util.SDK_INT >= 23

**Comment by erdemguven** [comment]
Yes, please switch to androidX.

**Comment by loki666** [comment]
should be good

**Comment by erdemguven** [comment]
Thanks

**Pull request merged**